onnx/defs/traditionalml/defs.cc

```diff
@@ -505,15 +515,72 @@ ONNX_ML_OPERATOR_SET_SCHEMA(
             std::string("NONE"))
         .TypeAndShapeInferenceFunction([](InferenceContext& ctx) {
           std::vector<std::string> label_strs;
-          auto result =
+          std::vector<int64_t> label_ints;
+
+          auto labels_strings_present =
               getRepeatedAttribute(ctx, "classlabels_strings", label_strs);
-          bool using_strings = (result && !label_strs.empty());
-          auto output_elem_type = ctx.getOutputType(0)->mutable_tensor_type();
+          bool using_strings = (labels_strings_present && !label_strs.empty());
+
+          if (!using_strings) {
+            getRepeatedAttribute(ctx, "classlabels_ints", label_ints);
+          }
+
+          // Type inference
+          auto* output_elem_type = ctx.getOutputType(0)->mutable_tensor_type();
           if (using_strings) {
             output_elem_type->set_elem_type(TensorProto::STRING);
           } else {
             output_elem_type->set_elem_type(TensorProto::INT64);
           }
+
+          // second output is always of float type
+          ctx.getOutputType(1)->mutable_tensor_type()->set_elem_type(
+              TensorProto::FLOAT);
+
+          // Shape/Rank inference begins
+
+          // establish the number of classes
+          std::vector<float> intercepts;
+          getRepeatedAttribute(ctx, "intercepts", intercepts);
+          int class_count = static_cast<int>(intercepts.size());
+          if (intercepts.size() == 1 &&
+              ((using_strings && label_strs.size() == 2) ||
+               (!using_strings && label_ints.size() == 2))) {
+            class_count = 2;
+          }
+
+          auto* output_shape_0 =
```

Comment:
This may be a subjective/personal preference, but I feel the following lines could be replaced as below. I perfonally find this a bit easier to read and check:
```cpp
    TensorShapeProto::Dimension batch_size_dim, class_count_dim;
    class_count_dim->set_dim_value(class_count);
    if (hasNInputShapes(ctx, 1)) {
        const auto& input_shape = ctx.getInputType(0)->tensor_type().shape();
        const auto input_rank = input_shape.dim_size();
        if (input_rank == 1) {
          // if input_rank is 1, batch_size is interpreted to be 1
          batch_size_dim->set_dim_value(1);
        } else if (input_rank == 2) {
          batch_size_dim = input_shape.dim((int)0);
        } else {
          fail_shape_inference("Input's shape should be 1D or 2D");
        }
    }
    updateOutputShape(ctx, 0, {batch_size_dim});
    updateOutputShape(ctx, 1, {batch_size_dim, class_count});
}
```

Reply:
Hi @gramalingam - that does look better. Thanks for your feedback. I will incorporate it.